crates/bevy_animation/src/curve_path.rs

```diff
@@ -0,0 +1,159 @@
+//! A curve through space.
+
+use bevy_math::cubic_splines::{CubicCurve, Point};
+
+/// A curve representing a path through space. Unlike [`CubicCurve`], it is sampled based on distance
+/// along the path. Can be used for linearly animating along a curve.
+///
+/// Note: Values are approximated using a series of line segments. Accuracy is based on the
+/// subdivisions specified when creating the path.
+#[derive(Clone, Debug)]
+pub struct CurvePath<P: Point> {
+    curve: CubicCurve<P>,
+    arc_lengths: Vec<f32>,
+}
+
+impl<P: Point> CurvePath<P> {
+    /// Creates a new [`CurvePath`] from a curve. Subdivisions will determine how many line segments
+    /// are used, and therefore the accuracy.
+    pub fn from_cubic_curve(curve: CubicCurve<P>, subdivisions: usize) -> CurvePath<P> {
+        let arc_lengths: Vec<f32> = curve
+            .iter_positions(subdivisions)
+            .scan((0.0, curve.position(0.0)), |state, x| {
+                state.0 += x.distance(state.1);
+                state.1 = x;
+
+                Some(state.0)
+            })
+            .collect();
+
+        CurvePath { curve, arc_lengths }
+    }
+
+    /// Returns the [`CubicCurve`] this path is based on.
+    pub fn curve(&self) -> &CubicCurve<P> {
+        &self.curve
+    }
+
+    /// Total length of the curve.
+    pub fn length(&self) -> f32 {
+        self.arc_lengths[self.arc_lengths.len() - 1]
+    }
+
+    /// Returns a 't' value corresponding to the `t` value on the underlying curve that matches the
+    /// provided `distance`.
+    pub fn t(&self, distance: f32) -> f32 {
+        // Get index with greatest value that is less than or equal to target distance.
+        let closest = self.arc_lengths.partition_point(|&x| x <= distance) - 1;
+
+        // Check if index's distance perfectly matches target distance, otherwise lerp between the
+        // index and its next neighbor.
+        if self.arc_lengths[closest] == distance {
+            (closest + 1) as f32 / (self.arc_lengths.len() + 1) as f32
+        } else {
+            let length_before = self.arc_lengths[closest];
+            let length_after = self.arc_lengths[closest + 1];
+            let segment_length = length_after - length_before;
+
+            let segment_fraction = (distance - length_before) / segment_length;
+
+            (closest as f32 + segment_fraction) / (self.arc_lengths.len() - 1) as f32
+                * self.curve.segments().len() as f32
+        }
+    }
+
+    /// Compute the position of a point on the curve at `distance`.
+    pub fn position(&self, distance: f32) -> P {
+        self.curve.position(self.t(distance))
+    }
+
+    /// Compute the first derivative at `distance`. This is the instantaneous velocity of a point on the cubic curve at `distance`.
+    pub fn velocity(&self, distance: f32) -> P {
```

Comment:
I sense a potential footgun with this method (and acceleration).
When working with raw beziers, `velocity(t)` returns the bezier-curve velocity, aka the derivative of the `position(t)` function wrt the argument `t`.

However, the `velocity(distance)` function does not return the derivative of `position(distance)` wrt  its argument `distance`, it is still the derivative according to the parameter `t`, which is not even visible in this API.

The reason a user might be confused by this would be for example when animating objects by monotonically increasing the `distance` and wanting to know their velocities and accelerations, however the methods provided here don't actually return that (the velocity would be incorrectly scaled (assuming nondegenerate beziers and numerically correct implementation, the magnitude of the velocity vector should be 1 if I'm not mistaken) and the acceleration method would likely yield completely unrecognizable values).

Ofc this is a hypothetical, but I could see this happening, explaining it in documentation could remedy this issue.